### PR TITLE
Add REST API for appointment configs

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/calendar/issues/3481

Base URL is `/apps/calendar/v1/appointment_configs`. You can `GET` it to get all configs, and `GET`, `POST` and `DELETE` a specific config at `/apps/calendar/v1/appointment_configs/123`.

### CODE REVIEW

File: appinfo/routes.php [hunk 1]
```diff
@@ -56,5 +56,8 @@
 		['name' => 'settings#setConfig', 'url' => '/v1/config/{key}', 'verb' => 'POST'],
 		// Tools
 		['name' => 'email#sendEmailPublicLink', 'url' => '/v1/public/sendmail', 'verb' => 'POST'],
+	],
+	'resources' => [
+		'appointmentConfig' => ['url' => 'api/appointment_configs']
```
Comment (tcitworld) [inline appinfo/routes.php:61]: No `/v1/` prefix?

Comment (ChristophWurst) [reply appinfo/routes.php:61]: It's an internal API. What's the purpose of the versioning? if the API changes we won't keep two versions anyway, right?

File: lib/Controller/AppointmentConfigController.php [hunk 2]
```diff
@@ -22,72 +23,67 @@
...
- *
- * @package OCA\Calendar\Controller
- */
 class AppointmentConfigController extends Controller {
 
-	/** @var IInitialStateService */
-	private $initialStateService;
-
-	/** @var IUser */
-	private $user;
```
Comment (st3iny) [inline lib/Controller/AppointmentConfigController.php:48]: The user should stay. Currently, the parameter `$userId` is allowed to be null in the constructor but later in the `create()` method it is required to be non-null -> might go :boom:

File: lib/Controller/AppointmentConfigController.php [hunk 3]
```diff
@@ -22,72 +23,67 @@
...
-								IInitialStateService     $initialStateService,
-								IUser                    $user,
-								AppointmentConfigService $appointmentService) {
+	/** @var string|null */
+	private $userId;
+
+	public function __construct(string $appName,
+								IRequest $request,
+								AppointmentConfigService $appointmentService,
+								?string $userId) {
```
Comment (st3iny) [inline lib/Controller/AppointmentConfigController.php:49]: Or we can forbid null for $userId here.

**Review by st3iny — COMMENTED** [review]
~Also see https://github.com/nextcloud/calendar/pull/3470#discussion_r741235847.~

~(Can't mark it here because it's not in the diff.)~

**EDIT:** resolved

File: lib/Controller/AppointmentConfigController.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/Controller/AppointmentConfigController.php:49]: doesn't work, then accessing the route anonymously triggers a type error. we first instantiate the controller before method permissions (`@PublicPage`) is checked

File: lib/Controller/AppointmentConfigController.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/Controller/AppointmentConfigController.php:48]: I will add a null check.

**Comment by ChristophWurst** [comment]
@st3iny creation was now tested with a good old `$.post('/apps/calendar/v1/appointment_configs', {'name': 'test', 'description': 'desc', 'location': 'loc', 'visibility': 'viz', 'targetCalendarUri': 'personal-1', 'availability': '', length: 40, increment: 10, preparationDuration: 0, followupDuration: 10, buffer: 10, dailyMax: null, freebusyUris: ['a']})`. This worked :shipit:

**Pull request merged**